### PR TITLE
Inv2.0/add support for arel union distinct

### PR DESCRIPTION
**Pull request by nickfarina**
This is a companion PR for https://github.com/blueapron/wms-server/pull/1696, where we'll be making changes to the way this gem generates SQL in order to coerce the postgres optimizer to behave in the way we want it to.

### CODE REVIEW

File: lib/arel/nodes/postgresql.rb [hunk 1]
```diff
@@ -43,6 +46,11 @@ def visit_Arel_Nodes_ArrayConcat o, collector
           visit o.right, collector
         end
       end
+
+      def visit_Arel_Nodes_UnionDistinct o, collector
```
Comment (nickfarina) [inline lib/arel/nodes/postgresql.rb]: This is unintuitive, but is the Arel way of handling this. Taken from https://github.com/rails/arel/blob/master/lib/arel/visitors/to_sql.rb#L302. We will likely be able to contribute this addition to the core project.

File: lib/active_record/hierarchical_query/cte/recursive_term.rb [hunk 2]
```diff
@@ -25,7 +25,11 @@ def arel
 
         private
         def scope
-          @scope ||= query.child_scope_value.select(columns)
+          @scope ||= if builder.options.key?(:seed_recursive_query) && builder.options[:seed_recursive_query] == true
+            query.start_with_value.select(columns)
+          else
+            query.child_scope_value.select(columns)
+          end
```
Comment (nickfarina) [inline lib/active_record/hierarchical_query/cte/recursive_term.rb]: Notes to self -
I'm wondering if there is any reason the recursive term is not always seeded with the same AR as the non recursive term. 

What is meant by `child_scope_value`?

Comment (keithmgould) [inline lib/active_record/hierarchical_query/cte/recursive_term.rb]: is this something you want to look into before we go ahead? (or look into...more?)

Comment (nickfarina) [inline lib/active_record/hierarchical_query/cte/recursive_term.rb]: yes, i'm looking into it now. i can't see a reason why you would want different queries on each side of the union, but maybe there is? will report back

**Comment by nickfarina** [comment]
Good news, no need to seed the recusive query as well, we can achieve the same effect by appending the `where` clauses we need _after_ the `connect_by` clause

File: lib/arel/nodes/postgresql.rb (hunk 1 above)
Comment (nickfarina) [inline lib/arel/nodes/postgresql.rb]: Here is a PR for the core arel project -> https://github.com/rails/arel/pull/442

I will leave these changes in here, and remove them later if my PR gets merged.

File: spec/active_record/hierarchical_query_spec.rb [hunk 3]
```diff
@@ -11,6 +11,33 @@
   let!(:child_5) { klass.create(parent: child_4) }
 
   describe '#join_recursive' do
+    describe 'UNION clause' do
+      let(:options) { {} }
+      subject { klass.join_recursive(options) { connect_by(id: :parent_id) }.to_sql }
+
+      it 'defaults to UNION ALL' do
+        expect(subject).to include('UNION ALL')
```
Comment (keithmgould) [inline spec/active_record/hierarchical_query_spec.rb]: this the two test cases below include an include and a not include, what about adding a "not include union distinct" here

**Comment by keithmgould** [comment]
minor comment, NB.

👍

**Comment by nickfarina** [comment]
Closing because these changes were merged into the base repo https://github.com/take-five/activerecord-hierarchical_query

**Pull request closed without merging**